File: webapp/backend/internal/models/banner.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// Banner represents a promotional banner displayed in the mobile app.
type Banner struct {
	ID        uuid.UUID      `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	Title     string         `gorm:"type:varchar(255);not null" json:"title"`
	Subtitle  string         `gorm:"type:text" json:"subtitle,omitempty"`
	ImageURL  string         `gorm:"type:varchar(500)" json:"image_url,omitempty"`
	IconURL   string         `gorm:"type:varchar(500)" json:"icon_url,omitempty"`
	IconKey   string         `gorm:"type:varchar(60)" json:"icon_key,omitempty"`
	CtaText   string         `gorm:"type:varchar(100)" json:"cta_text,omitempty"`
	CtaType   string         `gorm:"type:varchar(20);default:'route'" json:"cta_type,omitempty"`
	CtaRoute  string         `gorm:"type:varchar(255)" json:"cta_route,omitempty"`
	BgColor   string         `gorm:"type:varchar(20)" json:"bg_color,omitempty"`
	Locations []string       `gorm:"type:jsonb;serializer:json;default:'[\"home\"]'" json:"locations"`
	Platform  string         `gorm:"type:varchar(20);not null;default:'all'" json:"platform"`
	IsActive  bool           `gorm:"default:true" json:"is_active"`
	SortOrder int            `gorm:"default:0" json:"sort_order"`
	StartDate *time.Time     `gorm:"type:timestamptz" json:"start_date,omitempty"`
	EndDate   *time.Time     `gorm:"type:timestamptz" json:"end_date,omitempty"`
	CreatedAt time.Time      `gorm:"type:timestamptz;not null;default:now()" json:"created_at"`
	UpdatedAt time.Time      `gorm:"type:timestamptz;not null;default:now()" json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"type:timestamptz;index" json:"deleted_at,omitempty"`
}

func (Banner) TableName() string {
	return "banners"
}

File: webapp/frontend/src/types/banner.ts
// ============================================
// Banner Types
// ============================================

export type Platform = "all" | "android" | "ios";

export interface Banner {
  id: string;
  title: string;
  subtitle?: string;
  image_url?: string;
  icon_url?: string;
  icon_key?: string;
  cta_text?: string;
  cta_type?: "route" | "url";
  cta_route?: string;
  bg_color?: string;
  type: string;
  platform: Platform;
  is_active: boolean;
  sort_order: number;
  start_date?: string;
  end_date?: string;
  created_at: string;
  updated_at: string;
}

export interface CreateBannerRequest {
  title: string;
  subtitle?: string;
  image_url?: string;
  icon_url?: string;
  icon_key?: string;
  cta_text?: string;
  cta_type?: "route" | "url";
  cta_route?: string;
  bg_color?: string;
  type?: string;
  platform?: Platform;
  is_active?: boolean;
  sort_order?: number;
  start_date?: string;
  end_date?: string;
}

export interface UpdateBannerRequest {
  title?: string;
  subtitle?: string;
  image_url?: string;
  icon_url?: string;
  icon_key?: string;
  cta_text?: string;
  cta_type?: "route" | "url";
  cta_route?: string;
  bg_color?: string;
  type?: string;
  platform?: Platform;
  is_active?: boolean;
  sort_order?: number;
  start_date?: string;
  end_date?: string;
}

export interface BannerListParams {
  page?: number;
  limit?: number;
  search?: string;
}

export interface BannerOrder {
  id: string;
  sort_order: number;
}

export const PLATFORM_OPTIONS = [
  { value: "all", label: "Cả hai (Android & iOS)", short: "Cả hai" },
  { value: "android", label: "Chỉ Android", short: "Android" },
  { value: "ios", label: "Chỉ iOS", short: "iOS" },
] as const;

export const BANNER_TYPES = [
  { value: "feature", label: "Tính năng", color: "#BF360C" },
  { value: "content", label: "Bài viết", color: "#1565C0" },
  { value: "quiz", label: "Đố vui", color: "#2E7D32" },
  { value: "ai", label: "AI", color: "#4A148C" },
  { value: "promo", label: "Khuyến mãi", color: "#E65100" },
  { value: "custom", label: "Tùy chỉnh", color: "#37474F" },
] as const;
